### ISSUE TITLE
Set header height in headerStyle in native-stack?

### ISSUE THREAD

**Issue by gonzalo-rivas** [react-navigation/react-navigation#10097]
### Current behavior

I am migrating to react navigation 6 and I want to use the new standard native-stack library to upgrade the performance of these stacks. I have from react-navigation 5 this stack navigator (a snippet):

Screen1StackNavigator:
`  return (
    <Stack.Navigator
      screenOptions={{
        headerStyle: {
          height: 100,
          backgroundColor: 'black',
        },
      }}
    >
      <Stack.Screen
        name="Screen1"
        component={Screen1}
        options={{
          headerTitle: () => <Title title={'Screen1'} />,
        }}
      />
    </Stack.Navigator>
  )`

This snippet is in this tree of the main navigator:
MainTabsNavigator:

` <Tab.Navigator
      screenOptions={{
        lazy: true,
        headerShown: false,
      }}
      ...
    >
      <Tab.Screen
        name="Screen1"
        component={Screen1StackNavigator}
        options={{
          TabLabel: t('markets.title'),
        }}
      />...`

But with this native-stack, the header style cant set its height. I saw in the docs that the only props you can set are:

Style object for header. Supported properties:

backgroundColor

So I want to know if there is a hidden tricky way to set the height as I used to do in react navigation 5 with "default" stack.

### Expected behavior

Set the inner Stack Header Height to a specific height.

### Reproduction

-

### Platform

- [X] Android
- [X] iOS
- [ ] Web
- [ ] Windows
- [ ] MacOS

### Packages

- [ ] @react-navigation/bottom-tabs
- [ ] @react-navigation/drawer
- [ ] @react-navigation/material-bottom-tabs
- [ ] @react-navigation/material-top-tabs
- [ ] @react-navigation/stack
- [X] @react-navigation/native-stack

### Environment

"@react-navigation/bottom-tabs": "^6.0.9",
    "@react-navigation/drawer": "^6.1.8",
    "@react-navigation/elements": "^1.2.1",
    "@react-navigation/native": "^6.0.6",
    "@react-navigation/native-stack": "^6.2.5",
    "@react-navigation/stack": "^6.0.11",

**Comment by tylercote** [comment]
+1 I am also seeing this issue going from v5 stack to v6 native-stack.  Setting height on the headerStyle is no longer applying.

**Comment by matias-calvo-91** [comment]
I found a workaround for this issue.
If you use the `header` prop, you can set **height** specifying the whole component; like this:

```
<Stack.Screen
  options={
    header: () =>
        (
          <View style={{ height: 100 }}>
            <BackArrow navigation={navigation.goBack} style={{ position: 'absolute', left: 30, bottom: 35 }} />
            <View>
              <Text>TITLE</Text>
              <Text>TITLE</Text>
            </View>
          </View>
        ),
  }
/>
```

But you would need to add the back arrow manually.
It's not a solution, but at least it helped me.

**Comment by J6ey** [comment]
@matias-calvo-91 Thanks for sharing. It's a great solution, however, the header will go out of the status bar in iOS.

**Comment by liquidvisual** [comment]
@matias-calvo-91 Thanks! That worked for me. Do you know if it's possible to make the custom header absolute though? When I do so, it disppears.

**Comment by AshotN** [comment]
Setting the header height in the `header` prop doesn't automatically push the content down.

**Comment by Aure77** [comment]
> #10113 was supposed to fix this but setting height only works for web.

Only works if you use custom `header`.

**Comment by mcatalano54** [comment]
setting the "header" prop in options, the height changes correctly, but the "headerLeft" and "headerRight" props stopping working. I defined some icons there.  So if  I'm not wrong, I'll need to add that icons inside the custom view defined in the "header" option.

**Comment by chriscoomber** [comment]
To anyone stumbling across this, I managed to get a workaround that worked for me which I can share.

Whilst setting the `height` via `options.headerStyle` is not officially supported (only `backgroundColor` is mentioned in the docs), it does basically work. https://github.com/react-navigation/react-navigation/blob/5b13f818f3c701ce33c2a30f5dd96763069556e4/packages/elements/src/Header/Header.tsx#L86

But with the following gotcha:

The header height includes any insets due to the StatusBar, or any other notches. So, naively setting the header height to what you want it to be might cause issues on Android. If you want to set the header height to 100, you probably want to do something like:

```
import { useSafeAreaInsets } from 'react-native-safe-area-context';

...

const insets = useSafeAreaInsets();

...

<RootStack.Navigator
      initialRouteName="Home"
    >
      <RootStack.Screen
        name="Home"
        component={HomeScreen}
        options={{
          headerStyle: {
            height: 100 + insets.top
          }
        }}
      />

...
```

I wanted to scale the header height depending on the screen size, for a fluid UI (don't ask...), and for me it was helpful to know the header's default height (i.e. what it is if you don't set it). To calculate this you can do:

```
import { useWindowDimensions } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { getDefaultHeaderHeight } from '@react-navigation/elements';

...

const dimensions = useWindowDimensions();
const insets = useSafeAreaInsets();
const defaultHeaderHeight = getDefaultHeaderHeight(dimensions, false, insets.top);  // assuming your header never appears on a modal screen, if it does the middle argument should be true
const newHeaderHeight = myCustomScalingCode(defaultHeaderHeight - insets.top) + insets.top;
```

**Comment by Aure77** [comment]
https://github.com/react-navigation/react-navigation/blob/5b13f818f3c701ce33c2a30f5dd96763069556e4/packages/native-stack/src/views/NativeStackView.native.tsx#L177

Header height is managed here. It use default height when no custom header is provided...

**Comment by tsheaff** [comment]
Damn, am I understanding this correctly that *there is no way to change the height of the header* other than a completely custom from-scratch header component?

**Comment by endcycles** [comment]
Question: How can I change the height of the nav component in React Navigation 6 with the new native-stack library?

Current behavior:
I am currently exploring React Navigation 6 and using the native-stack library for my navigation. I have created a stack navigator with a header style that includes a specific height using the headerStyle property. However, the height property doesn't seem to have any effect on the header in the new native-stack. Is there a way to set the height of the header in React Navigation 6 with native-stack?

Expected behavior:
I want to set the height of the inner Stack header to a specific value.

Platform:
iOS
MacOS

Packages:
@react-navigation/native-stack
@react-navigation/native

**Comment by bryanltobing** [comment]
> Damn, am I understanding this correctly that there is no way to change the height of the header other than a completely custom from-scratch header component?

if this is the case.

Is there a guide on how to properly build a custom header? if i make my own implementation i always can't get it right with the transition, and the title

**Comment by KCErb** [comment]
In my case, I wrapped my app with "safe area insets" and found that my headers were too tall because they were including the status bar height.

So I tried using the `Header` component from the Elements library so that I could set the `headerStatusBarHeight` property: https://reactnavigation.org/docs/elements/#headerstatusbarheight

To my surprise, before I actually called the component, the header bar height was fixed! So I uninstalled the elements library and the height went back to being too tall.

### TL;DR

Try

```yarn add @react-navigation/elements```

it might just make your header the height you want as a side effect.

**Comment by KCErb** [comment]
Update: your mileage may vary on the above. I'm already having trouble reproducing consistently.

**Comment by michaelkremenetsky** [comment]
Can this be reopened because of https://x.com/swmansion/status/1714324990491762873?s=46

**Comment by buddywood** [comment]
I had to use a custom Header component to get it to work.

**Comment by dgrcode** [comment]
> Try
> 
> yarn add @react-navigation/elements
> 
> and then use the Header component, even if you get rid of it later, merely importing the library seems to have this side effect.

I've tried this, but I lose the back button and any `headerRight` I had defined. Am I missing something?

**Comment by anyamiletic** [comment]
@dgrcode if you use a custom `header` component it actually replaces `headerLeft`, `headerRight` and `headerTitle` as a single component. Your UI will no longer render those and so you will have to render the whole header UI in your custom component

**Comment by timbtimbtimb** [comment]
I'd love to have control over the height too.

**Comment by sirusbaladi** [comment]
why is this closed?

**Comment by ameer2468** [comment]
Experiencing the same issue, need to pass a custom header component instead which is annoying.

**Comment by shadeemerhi** [comment]
Experiencing the same issue and it's extremely annoying

**Comment by DevOkone** [comment]
header: () =><View style={{ height: 50, backgroundColor:"red" }}  />

**Comment by mo92othman** [comment]
Still no solution apart from using a custom header component?

**Comment by fredrikburmester** [comment]
i'd also like a solution that's not a custom header component.

**Comment by Steffi3rd** [comment]
With Expo Router, you can use `header` option with the `<Header />` component and add the props `headerStyle`

It will override only the `headerStyle` props without breaking`headerLeft`, `headerTitle`...

```
import { Header } from "@react-navigation/elements";
import { Stack } from "expo-router";

    <Stack
      screenOptions={{
          header: ({ options }) => (
            <Header
              {...options}

              headerStyle={{
                height: 100,
                backgroundColor: "red",
              }}
            />
          ),
        }),
      }}
    >
    enter code here
    </Stack>
```

Source: https://reactnavigation.org/docs/elements/#header

**Comment by kdechant** [comment]
For Expo Router, another workaround for setting the height is to switch to @react-navigation/stack, as described here:

https://docs.expo.dev/router/advanced/stack/#javascript-stack-with-react-navigationstack

**Comment by itaikta** [comment]
Adding on to [Steffi3rd](https://github.com/Steffi3rd), to get the back button dynamically:
```
header: ({ options, back }) => (
    <Header
        {...options}
         back={back}
    />
)
```
